corehq/blobs/s3db.py

```diff
@@ -68,17 +69,25 @@ def put(self, content, **blob_meta_args):
                 s3_bucket.copy(source, meta.key)
         else:
             content.seek(0)
-            meta.content_length = get_file_size(content)
+            if meta.compressed:
+                meta.content_length = get_file_size(GzipCompressReadStream(content))
+                # Need to re-initialize stream as cannot seek on Gzipstream
+                content.seek(0)
+                content = GzipCompressReadStream(content)
```

Comment:
can we avoid doing the compression twice?

```suggestion
            if meta.compressed:
                buf = io.BytesIO()
                with GzipFile(fileobj=buf, mode='wb') as f:
                    f.write(data)
                content = buf
            meta.content_length = get_file_size(content)
```
I realize that this will load all the data in memory but that's probably OK for our current use cases (at least it would simplify things initially and we could optimize if we need to compress larger objects). If we did this we could get rid of `GzipCompressReadStream` completely

Reply:
@snopoke I would prefer to maintain support for arbitrarily large files. The blob db has been designed that way from the beginning, and I think it's an important part of the mental model for building things on top of it.

My recommendation for accomplishing that would be to have `GzipCompressReadStream` keep track of how many bytes have been read
```py
            if meta.compressed:
                content = GzipCompressReadStream(content)
            with self.report_timing('put', meta.key):
                s3_bucket.upload_fileobj(content, meta.key)
            meta.content_length = get_file_size(content)
            self.metadb.put(meta)
        return meta

...

def get_file_size(fileobj):
    if isinstance(fileobj, GzipCompressReadStream):
        return fileobj.content_length
```
`<GzipCompressReadStream>.content_length` implementation TBD. The attribute should not be set initially (it is an error to read the content length before all bytes have been read). It should be set to an integer value once all bytes have been read from the input.

Caveat: I can't remember if it's important to save the metadata _before_ writing blob content to S3. If it is then we have a big problem since we could never optimize this for arbitrary file size.

However, there is another question I'm not sure we've considered: is `meta.content_length` the number of bytes stored in the backend, or is it the number of bytes that would be returned by reading the file-like object returned by `blobdb.get(...)`? I can see reasons why we would want both, and it seems possibly important to have it be the latter since the blob db handles compression transparently. Thoughts?

Reply:
> However, there is another question I'm not sure we've considered: is meta.content_length the number of bytes stored in the backend (compressed bytes), or is it the number of bytes that would be returned by reading the file-like object returned by blobdb.get(...) (uncompressed bytes)? I can see reasons why we would want both, and it seems possibly important and more intuitive to have it be the latter since the blob db handles compression transparently. Thoughts?

I was wondering about this too. We don't really use this field very much but in the few places where it is used we need the uncompressed size (except where we post metrics to Datadog where we'd want the compressed size). One place were we wouldn't be able to get the uncompressed size is where we copy an existing blob: https://github.com/dimagi/commcare-hq/blob/8f63335b220751b0bdaac9e13abb818e0ab02413/corehq/blobs/s3db.py#L70. This appears to have been added specifically for case attachments here: https://github.com/dimagi/commcare-hq/pull/13203. I couldn't find any usages of it but there must be some since we're recording metrics for it on [production](https://app.datadoghq.com/dashboard/753-35q-fwt/blob-db-dashboard?from_ts=1584428385517&fullscreen_section=overview&fullscreen_widget=3073215485582670&live=true&tile_size=m&to_ts=1587020385517&tpl_var_env=production&utc_override=true&fullscreen_start_ts=1584446132694&fullscreen_end_ts=1587038132694&fullscreen_paused=false). Either way that does seem like a useful feature to keep. One approach would be to force the user to pass in the `meta` to `db.put` if they are passing in an existing blob or else pass the content length info into `BlobStream`.

I also noticed that if we return a `GzipFile` from `db.get` then the type check `isinstance(content, BlobStream)` will fail and we'll end up to a read then write.

Reply:
@millerdev I've made some updates here so that `meta.content_length` is the uncompressed size but there are 2 remaining challenges:

1. `db.size(key) != meta.content_length`

This is used in a few places to to add content length headers to responses. I think this could be solved by applying the same arg rules to `size` as we have to `get`. We could also eliminate all calls to `size` if we return the content length along with the blob from `get` since they are all in places where the actual blob is fetched as well.

2. The metrics are reporting the uncompressed size which I think is incorrect.

Not clear how to solve this since many places would not be able to get the compressed size easily e.g. bulk delete

Reply:
Have you considered adding a `compressed_size` attribute to `BlobMeta`? It would be null if `<BlobMeta>.compressed` is false. We could even improve storage efficiency a bit by removing `compressed` entirely and make the rule be that if `compressed_size` is null then the blob is not compressed, otherwise it is compressed. Thoughts?

Storing both `compressed_size` in addition to `content_length` is beginning to seem more straightforward than other options.